controller_manager:
  ros__parameters:
    update_rate: 500
    joint_state_broadcaster:
      type: joint_state_broadcaster/JointStateBroadcaster

    io_and_status_controller:
      type: ur_controllers/GPIOController

    speed_scaling_state_broadcaster:
      type: ur_controllers/SpeedScalingStateBroadcaster

    force_torque_sensor_broadcaster:
      type: force_torque_sensor_broadcaster/ForceTorqueSensorBroadcaster

    joint_trajectory_controller:
      type: joint_trajectory_controller/JointTrajectoryController

    scaled_joint_trajectory_controller:
      type: ur_controllers/ScaledJointTrajectoryController

    forward_velocity_controller:
      type: velocity_controllers/JointGroupVelocityController

    forward_position_controller:
      type: position_controllers/JointGroupPositionController

    cartesian_compliance_controller:
      type: cartesian_compliance_controller/CartesianComplianceController

    motion_control_handle:
      type: cartesian_controller_handles/MotionControlHandle

    cartesian_adaptive_compliance_controller:
      type: cartesian_adaptive_compliance_controller/CartesianAdaptiveComplianceController

motion_control_handle:
  ros__parameters:
    end_effector_link: "wrist_3_link "
    robot_base_link: "world"
    ft_sensor_ref_link: "wrist_3_link"
    joints:
      - shoulder_pan_joint
      - shoulder_lift_joint
      - elbow_joint
      - wrist_1_joint
      - wrist_2_joint
      - wrist_3_joint


motion_control_handle:
  ros__parameters:
    end_effector_link: "wrist_3_link"
    robot_base_link: "world"
    ft_sensor_ref_link: "wrist_3_link"
    joints:
      - shoulder_pan_joint
      - shoulder_lift_joint
      - elbow_joint
      - wrist_1_joint
      - wrist_2_joint
      - wrist_3_joint


cartesian_adaptive_compliance_controller:
  ros__parameters:
    end_effector_link: "wrist_3_link"
    robot_base_link: "world"
    ft_sensor_ref_link: "wrist_3_link"
    compliance_ref_link: "wrist_3_link"
    # end_effector_link: "wrist_3_link"
    # robot_base_link: "base_link"
    # ft_sensor_ref_link: "wrist_3_link"
    # compliance_ref_link: "wrist_3_link"
    joints:
      - shoulder_pan_joint
      - shoulder_lift_joint
      - elbow_joint
      - wrist_1_joint
      - wrist_2_joint
      - wrist_3_joint

    command_interfaces:
      - position

    stiffness: 
        trans_x: 500.0
        trans_y: 500.0
        trans_z: 500.0
        rot_x: 20.0
        rot_y: 20.0
        rot_z: 20.0

    solver:
        error_scale: 0.5
        iterations: 1

    pd_gains:
        trans_x: {p: 0.05, d: 0.005}
        trans_y: {p: 0.05, d: 0.005}
        trans_z: {p: 0.05, d: 0.005}
        rot_x: {p: 1.5}
        rot_y: {p: 1.5}
        rot_z: {p: 1.5}

    # Parameters of the QP and adaptive controller
 # Topics
    target_velocity_topic: "/target_velocity"
    target_wrench_topic: "/target_wrench"

    # Sampling period of the interface
    sampling_period: 0.002

    # Tank parameters
    tank:
        initial_state: 1.0
        minimum_energy: 0.4
        eta: -0.1

    # QP parameters
    Qp:
        F_min:  # Force lower bound
            - -15.0
            - -15.0
            - -15.0
            - -1.5
            - -1.5
            - -1.5
        F_max:  # Force upper bound
            - 15.0
            - 15.0
            - 15.0
            - 1.5
            - 1.5
            - 1.5
        K_min: # Minimum stiffness (diagonal)
            - 0.1
            - 0.1
            - 0.1
            - -0.10
            - -0.10
            - -0.10
        K_max: # Maximum stiffness (diagonal)
            - 1000.0
            - 1000.0
            - 1000.0
            - 100.0
            - 100.0
            - 100.0
        Q_weights: # Q weights
            - 3200.0
            - 3200.0
            - 3200.0
            - 3200.0
            - 3200.0
            - 3200.0
        R_weights: # R weights
            - 0.01
            - 0.01
            - 0.01
            - 0.01
            - 0.01
            - 0.01

speed_scaling_state_broadcaster:
  ros__parameters:
    state_publish_rate: 100.0


force_torque_sensor_broadcaster:
  ros__parameters:
    sensor_name: tcp_fts_sensor
    state_interface_names:
      - force.x
      - force.y
      - force.z
      - torque.x
      - torque.y
      - torque.z
    frame_id: tool0
    topic_name: ft_data


joint_trajectory_controller:
  ros__parameters:
    joints:
      - shoulder_pan_joint
      - shoulder_lift_joint
      - elbow_joint
      - wrist_1_joint
      - wrist_2_joint
      - wrist_3_joint
    command_interfaces:
      - position
    state_interfaces:
      - position
      - velocity
    state_publish_rate: 100.0
    action_monitor_rate: 20.0
    allow_partial_joints_goal: false
    constraints:
      stopped_velocity_tolerance: 0.2
      goal_time: 0.0
      shoulder_pan_joint: { trajectory: 0.2, goal: 0.1 }
      shoulder_lift_joint: { trajectory: 0.2, goal: 0.1 }
      elbow_joint: { trajectory: 0.2, goal: 0.1 }
      wrist_1_joint: { trajectory: 0.2, goal: 0.1 }
      wrist_2_joint: { trajectory: 0.2, goal: 0.1 }
      wrist_3_joint: { trajectory: 0.2, goal: 0.1 }


scaled_joint_trajectory_controller:
  ros__parameters:
    joints:
      - shoulder_pan_joint
      - shoulder_lift_joint
      - elbow_joint
      - wrist_1_joint
      - wrist_2_joint
      - wrist_3_joint
    command_interfaces:
      - position
    state_interfaces:
      - position
      - velocity
    state_publish_rate: 100.0
    action_monitor_rate: 20.0
    allow_partial_joints_goal: false
    constraints:
      stopped_velocity_tolerance: 0.2
      goal_time: 0.0
      shoulder_pan_joint: { trajectory: 0.2, goal: 0.1 }
      shoulder_lift_joint: { trajectory: 0.2, goal: 0.1 }
      elbow_joint: { trajectory: 0.2, goal: 0.1 }
      wrist_1_joint: { trajectory: 0.2, goal: 0.1 }
      wrist_2_joint: { trajectory: 0.2, goal: 0.1 }
      wrist_3_joint: { trajectory: 0.2, goal: 0.1 }

forward_velocity_controller:
  ros__parameters:
    joints:
      - shoulder_pan_joint
      - shoulder_lift_joint
      - elbow_joint
      - wrist_1_joint
      - wrist_2_joint
      - wrist_3_joint
    interface_name: velocity

forward_position_controller:
  ros__parameters:
    joints:
      - shoulder_pan_joint
      - shoulder_lift_joint
      - elbow_joint
      - wrist_1_joint
      - wrist_2_joint
      - wrist_3_joint
cartesian_compliance_controller:
  ros__parameters:
    end_effector_link: "tool0"
    robot_base_link: "world"
    ft_sensor_ref_link: "tool0"
    compliance_ref_link: "tool0"

    joints:
      - $(var tf_prefix)shoulder_pan_joint
      - $(var tf_prefix)shoulder_lift_joint
      - $(var tf_prefix)elbow_joint
      - $(var tf_prefix)wrist_1_joint
      - $(var tf_prefix)wrist_2_joint
      - $(var tf_prefix)wrist_3_joint
      
    command_interfaces:
      - position

    stiffness:  # w.r.t. compliance_ref_link coordinates
        trans_x: 400.0
        trans_y: 400.0
        trans_z: 1000.0
        rot_x: 10.0
        rot_y: 10.0
        rot_z: 10.0

    solver:
        error_scale: 0.5
        iterations: 1

    # For all controllers, these gains are w.r.t. the robot_base_link coordinates.
    pd_gains:
        trans_x: {p: 0.05, d: 0.005}
        trans_y: {p: 0.05, d: 0.005}
        trans_z: {p: 0.3, d: 0.01}
        rot_x: {p: 1.5}
        rot_y: {p: 1.5}
        rot_z: {p: 1.5}

